### DISCUSSION TITLE
Easier process for testing data/metadata changes locally

### DISCUSSION THREAD

**Discussion by brockfanning** [open-sdg/open-sdg#1428] [Ideas]
Currently it's a little problematic to test data/metadata changes, since they have to normally be committed and pushed in order to get them built and deployed. If anyone has tips and tricks on this, please share!

Here is my own hack for this, which I think we could probably make less hacky:

Note: This requires Python, NodeJS and the [live-server](https://www.npmjs.com/package/live-server) tool: `npm install -g live-server`

1. Go to the data repo locally and make your changes to the data/metadata
2. Run the Python script to build the data: `python scripts/build_data.py`
3. Go into the generated folder: `cd _site`
4. Run live-server with the "cors" option: `live-server --cors`
5. Note the URL of the local site (eg, something like http://127.0.0.1:8000)
6. Go to the site repo
7. Edit the _config.yml to point to the live-server URL (noted above) instead of the remotely hosted data. Eg, change `remote_data_prefix: "https://my-github-org.github.io/open-sdg-data-starter"` to `remote_data_prefix: "http://127.0.0.1:8000"`.
8. Run the jekyll build and test locally: `bundle exec jekyll serve`

Problems:

* A lot of manual steps
* The data repo is not "watched" for changes, so you have to repeat the steps to make another change

Possible improvements:

* A "Makefile" or similar for the data repo to automate some of this, preferably with "watching" functionality
* A `_config_local.yml` file so you don't have to edit the _config.yml file

**Comment by dougmet** [comment]
Hi @brockfanning , yes agree this is tricky. In fact it's the main reason we built the test server to build/deploy all the branches.

A really neat solution, and something I know @SavvasStephanides is excited about, would be to run everything in Docker containers. With docker compose you can get each service in its own container, but then easily network them together in an isolated network. When you bring each service "up" it would trigger their build.

* Each project (including translations) would have a `Dockerfile` and an associated docker image (which has all the system dependencies installed).
* A project-wide docker-compose file would orchestrate bringing up each container, and running their builds.

Big advantages are that you setup each image once and it can all be automated.

Main disadvantage is the one-off setup cost.

As an aside: You can at least remove the node dependency above as python has an http server https://docs.python.org/3/library/http.server.html (not sure about the cors bit mind).

**Comment by dougmet** [comment]
The check part of sdg-build can at least be run to check the input data is valid. It's quite strict and will need some work when all the extra sdmx features arrive.

**Comment by brockfanning** [comment]
Great points @dougmet. I wonder if a quick fix might be a Python script that launches a web server and "watches" the codebase for changes. Like a homemade"jekyll serve" for the data repo.

@SavvasStephanides the Docker idea sounds amazing. I think our quick-start is still too painful, so if we could replace it with a few Docker commands, life would be good. I don't have much experience with Docker, but I'll start ramping up.

I imagine this will be in a new repo, like open-sdg-docker?

**Comment by brockfanning** [comment]
Related: #197

**Comment by LucyGwilliamAdmin** [comment]
The main audience for this is developers and platform maintainers who want to test changes locally - there is no user need right now so locking conversation
